src/state/safety.rs

```diff
@@ -0,0 +1,315 @@
+//! via nudges have consecutive views. If this flow is interrupted by a temporary loss of synchrony or
+//! a Byzantine leader, the validator-set-updating block has to be re-proposed, as specified in
+//! [repropose_block]. This rule is enforced through [safe_nudge]. Unlike block insertions, nudges
+//! cannot cause state updates unless they satisfy the rule.
+//! 
+//! ## Locking rules
+//! 
+//! Locking serves to guarantee the safety of commit. In SMR protocols that follow the lock-commit
+//! paradigm, a value has to be locked before it is committed. The idea is that even if only some
+//! replicas commit, the others hold a lock to the committed value and should eventually commit it.
+//! 
+//! The locked value is essentially a block, but we consider any QC that references the block
+//! to be valid proxy for it, and hence store a lockedQC instead. This enables keeping track
+//! of information on the view in which a quorum of replicas last voted for this block.
+//! 
+//! As mentioned above, if a block is locked, then any new proposal or nudge must "extend" it,
+//! unless there exists evidence that a quorum of replicas do not share this lock and have
+//! moved to an alternative branch.
+//! 
+//! In the pipelined HotStuff protocol, a block is locked if there is a 2-chain supporting it, i.e., 
+//! there is a QC for the block and a QC for its child, which serve as the block's prepare QC and
+//! precommit QC respectively. In such case it is the prepare QC that is locked. To adapt this rule to
+//! the non-pipelined protocol for validator-set-updating blocks, we lock the block on seeing a nudge
+//! with a precommitQC for it - it is the precommitQC that becomes locked this time, as by the 
+//! consecutive views requirement this is equivalent to locking the prepareQC. Since the block sync 
+//! protocol does not involve sending nudges, we must also lock the decideQC when referenced by the
+//! block. This is to ensure that subsequently received blocks are not accepted if they conflict with 
+//! it. This is enforced through [qc_to_lock].
+
+use crate::hotstuff::{
+    messages::Nudge, 
+    types::{Phase, QuorumCertificate}
+};
+use crate::types::{
+    basic::{ChainID, CryptoHash, ViewNumber}, 
+    block::Block
+};
+
+use super::{
+    block_tree::{BlockTree, BlockTreeError},
+    kv_store::KVStore
+};
+
+/// Returns whether a block can be considered safe, and thus cause updates to the block tree. 
+/// For this, it is necessary that:
+/// 1. safe_qc(&block.justify, block_tree, chain_id).
+/// 2. Its qc's must be either a generic qc or a decide qc.
+/// 
+/// Note that before inserting to the block tree, the caller should also check if
+/// no block with the same block hash is already in the block tree.
+///
+/// This function evaluates [safe_qc], then checks 2.
+///
+/// # Precondition
+/// [Block::is_correct]
+pub fn safe_block<K: KVStore>(block: &Block, block_tree: &BlockTree<K>, chain_id: ChainID) -> Result<bool, BlockTreeError> {
```

Comment:
Why is this function (as well as `safe_qc` and `safe_nudge`) public?

Reply:
I thought since they are safety predicates and have no active behaviour we could expose them to the user, but you're right I don't think it's necessary

Reply:
Yes, I think we should be consistent with the visibility of the definitions in this module. `qc_to_lock`, etc. are `pub(crate)`, so it’s a bit confusing that `safe_block`, `safe_qc`, and `safe_nudge` are public. 

Either they should be all public for a clear reason, or they should all be private (my preference).

Reply:
Makes sense, but you mean `pub(crate)`, not private, right? I made them all `pub(crate)`.

Reply:
@lyulka more generally, what are the things that we'd want to make public (other than the obvious "entry points" like `Replica` and `ReplicaSpec`)? I feel like I am not being consistent in other modules too.